utils/fixed_decimal/src/decimal.rs

```diff
@@ -350,65 +350,106 @@ impl FixedDecimal {
     /// let mut dec = FixedDecimal::from(42);
     /// assert_eq!("42", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(2);
+    /// dec.padded_left(4);
     /// assert_eq!("0042", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(-2);
-    /// assert_eq!("42", dec.to_string());    
+    /// dec.padded_left(3);
+    /// assert_eq!("042", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(-1);
-    /// assert_eq!("2", dec.to_string());
+    /// dec.padded_left(2);
+    /// assert_eq!("42", dec.to_string());
+    ///
+    /// dec.padded_left(1);
+    /// assert_eq!("42", dec.to_string());
     /// ```
-    pub fn pad_or_truncate_left(&mut self, shift: i16) {
-        self.upper_magnitude = match shift >= 0 {
-            true => self.upper_magnitude.checked_add(shift).unwrap_or(i16::MAX),
-            false => self.upper_magnitude.checked_add(shift).unwrap_or(0),
-        };
-
-        // upper_magnitude must be greater than or equal zero.
-        if self.upper_magnitude < 0 {
-            self.upper_magnitude = 0;
+    pub fn padded_left(&mut self, digits: u16) {
+        let mut magnitude = digits as i16 - 1;
+        if magnitude < 0 {
+            magnitude = 0;
         }
-
-        if self.upper_magnitude >= self.magnitude {
-            return;
+        // Do not truncate nonzero digits
+        if magnitude <= self.magnitude {
+            magnitude = self.magnitude;
         }
 
-        // how many digits need to be deleted from the start.
-        let cut = self.magnitude - self.upper_magnitude;
+        self.upper_magnitude = magnitude;
+    }
 
-        self.magnitude = self.upper_magnitude;
-        if cut >= self.digits.len() as i16 {
-            self.digits.clear();
-            self.magnitude = 0;
-            return;
+    /// Truncate the number on the left to a particular magnitude, deleting
+    /// digits if necessary.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fixed_decimal::FixedDecimal;
+    ///
+    /// let mut dec = FixedDecimal::from(4235);
+    /// assert_eq!("4235", dec.to_string());
+    ///
+    /// dec.truncated_left(5);
+    /// assert_eq!("4235", dec.to_string());
+    ///
+    /// dec.truncated_left(2);
+    /// assert_eq!("235", dec.to_string());
+    ///
+    /// dec.truncated_left(1);
+    /// assert_eq!("35", dec.to_string());
+    ///
+    /// dec.truncated_left(0);
+    /// assert_eq!("5", dec.to_string());
+    ///
+    /// dec.truncated_left(-1);
+    /// assert_eq!("0", dec.to_string());
+    /// ```
+    pub fn truncated_left(&mut self, magnitude: i16) {
+        if self.magnitude >= magnitude {
+            let positive_magnitude = if magnitude > 0 { magnitude } else { 0 };
+            let cut = self.magnitude - magnitude;
+            if cut >= self.digits.len() as i16 {
```

Comment:
Issue: I've been trying as much as possible to make FixedDecimal robust against integer overflow.  In this case, it may be necessary to cut more than (2^15-1) digits, but we cannot represent that since `cut` has type `i16`.  This can happen on a number with a large positive magnitude being truncated to a large negative magnitude.  Since `cut` needs to be a `usize` anyway, please change this to the equivalent of

```rust
let cut = ((self.magnitude as i32) - (magnitude as i32)) as usize;
```